modal_proto/sandbox_router.proto

```diff
@@ -0,0 +1,126 @@
+  // requests and ensure exec commands are idempotent.
+  //
+  // TODO(saltzm): Could instead have a separate idempotency key from the exec_id
+  // like present day, and have the server generate the exec_id and return it in
+  // the ExecStartResponse.
+  string exec_id = 2;
+  // Command arguments to execute.
+  repeated string command_args= 3;
+  // Configures how the stdout of the command will be handled.
+  SandboxExecStdoutConfig stdout_config = 4;
+  // Configures how the stderr of the command will be handled.
+  SandboxExecStderrConfig stderr_config = 5;
+  // Deadline for the exec'd command to exit. If the command does not exit before
+  // the deadline, the command will be killed. This is NOT the deadline for the
+  // ExecStartRequest RPC to complete.
+  google.protobuf.Timestamp deadline = 6;
+  // Working directory for the command.
+  optional string workdir = 7;
+  // Secret IDs to mount into the sandbox.
+  repeated string secret_ids = 8;
+  // PTY info for the command.
+  optional modal.client.PTYInfo pty_info = 9;
+  // Enable debugging capabilities on the container runtime. Used only for
+  // internal debugging.
+  bool runtime_debug = 10;
+}
+
+message SandboxExecStartResponse { }
+
+message SandboxExecStdinWriteRequest {
+  // The task ID of the sandbox running the exec'd command.
+  string task_id = 1;
+  // The execution ID of the command to write to.
+  string exec_id = 2;
+  // The offset to start writing to. This is used to resume writing from the
+  // last write position if the connection is closed and reopened.
+  uint64 offset = 3;
+  bytes data = 4;
+  // If true, close the stdin stream after writing any provided data.
+  // This signals EOF to the exec'd process.
+  bool eof = 5;
+}
+
+message SandboxExecStdinWriteResponse { }
+
+
+message SandboxExecStdioReadRequest {
+  // The task ID of the sandbox running the exec'd command.
+  string task_id = 1;
+  // The execution ID of the command to read from.
+  string exec_id = 2;
+  // The offset to start reading from. This is used to resume reading from the
+  // last read position if the connection is closed and reopened.
+  uint64 offset = 3;
+  // Which file descriptor to read from.
+  SandboxExecStdioFileDescriptor file_descriptor = 4;
+}
+
+message SandboxExecStdioReadResponse {
+  // The data read from the file descriptor.
+  bytes data = 1;
+}
+
+message SandboxExecWaitRequest {
+  // The task ID of the sandbox running the exec'd command.
+  string task_id = 1;
+  // The execution ID of the command to wait on. 
+  string exec_id = 2;
+}
+
+message SandboxExecWaitResponse {
+  oneof exit_status {
+    // The exit code of the command.
+    int32 code = 3;
+    // The signal that terminated the command.
+    int32 signal = 4;
+  }
+  // TODO(saltzm): Give a way for the user to distinguish between normal exit
+  // and termination by Modal (due to sandbox idle timeout, exec exceeded deadline, etc.)  
```

Comment:
FWIW, if there's an exec in progress, the sandbox should not hit the idle timeout. Agree that giving more detail about how an exec terminated is good, though!

Reply:
Ah true, fixed comment